### ISSUE TITLE
[BUG][SDW][CML] clock stop prepare issue - Speaker cannot work after suspend/resume

### ISSUE THREAD

**Issue by bardliao** [thesofproject/linux#2606]
Canonical reported an issue that speaker can't work on CML + dual rt1308 laptop.
The issue happens only once so far.
The error kernel log is
```
rt1308 sdw:1:25d:1308:0: Clock Stop prepare failed for slave: -61
rt1308 sdw:1:25d:1308:0: pre-prepare failed:-61
intel-master sdw-master-1: prepare clock stop failed -61
intel-master sdw-master-1: cannot enable clock stop on suspend
...
rt1308 sdw:1:25d:1308:0: Unable to sync register 0xc030. -22
intel-master sdw-master-1: pm_runtime_get_sync failed in intel_startup, ret -22
intel-master sdw-master-1: ASoC: can't open DAI SDW1 Pin2: -22
SDW1-Playback: ASoC: BE open failed -22
SDW1-speakers: ASoC: failed to start some BEs -22
```
The issue is that somehow we didn't get ACK from codec and then cause clock stop prepare failed.
The question is that do we have a way to recover from no ack received?

**Comment by plbossart** [comment]
can we have a longer dmesg trace @bardliao ?

What we could do in this case is just to mark the device as not functional, suspend anyways and reset the bus on startup.

But that looks like a corner case for now.

**Comment by mengdonglin** [comment]
@bardliao @plbossart It seems this issue can be reproduced with SDCA codec ALC1316 on TGL-U laptop device. Please check CI report 1118 on 2020-12-10 (UTC). Failed case: check-suspend-resume-with-playback-25.sh

dmesg
```
kernel: [   61.208533] PM: suspend exit
kernel: [   61.683079] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
kernel: [   61.683106] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
kernel: [   62.183129] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
kernel: [   62.183153] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
kernel: [   64.876111] intel-sdw intel-sdw.2: Msg ignored for Slave 1
kernel: [   64.876120] rt1316-sdca sdw:2:25d:1316:1: Clock Stop prepare failed for slave: -61
kernel: [   64.876130] rt1316-sdca sdw:2:25d:1316:1: pre-prepare failed:-61
kernel: [   64.876135] intel-sdw intel-sdw.2: prepare clock stop failed -61
kernel: [   64.876141] intel-sdw intel-sdw.2: cannot enable clock stop on suspend
```

**Comment by mengdonglin** [comment]
@plbossart @bardliao Can we take this issue as a duplicate of #2621 ?

**Comment by plbossart** [comment]
@mengdonglin please keep issues separate. it's not useful if we have the same bugs for CML+SoundWire 1.1 devices and TGL+SDCA. We don't know if the issues are the same or not. For now let's have a different bug for each sighting, thank you.

**Comment by bardliao** [comment]
@shumingfan said that there are only three scenarios that codec will not ack to the command.
1. Device number is wrong
2. Codec is powered down
3. Trying to write a read only register.

We have checked device number by https://github.com/thesofproject/linux/pull/2634 and the device number is correct. And we don't think it is case 3 unless there is something wrong in transportation. So the highest possibility is case 2.

**Comment by plbossart** [comment]
@bardliao I wonder if there's something in the clock stop prepare that turns something off, and it's not restored on resume?
Alternatively, what can happen is that the codec loses sync on clock stop, and that would cause the clock stop to fail.

**Comment by bardliao** [comment]
> @bardliao I wonder if there's something in the clock stop prepare that turns something off, and it's not restored on resume?
> Alternatively, what can happen is that the codec loses sync on clock stop, and that would cause the clock stop to fail.

@plbossart But shouldn't we get `-ETIMEDOUT` instead of `-ENODATA` if the codec loses sync?

**Comment by plbossart** [comment]
@bardliao the 'COMMAND_IGNORED' reply is provided immediately, at the end of the 48-bit command word. It's really a 'no reply' and it's clearly not deferred as a -ETIMEDOUT hint at.

**Comment by bardliao** [comment]
I tried to read `SDW_SCP_STAT` (0x44) in some places to check when codec starts to not ack. Based on the test result now, we can read `SDW_SCP_STAT` properly at the beginning of `intel_suspend_runtime()`, and `-ENODATA` can be seen before `sdw_bus_prep_clk_stop` call `sdw_slave_clk_stop_callback`.  So I think we lost codec in the `sdw_cdns_clock_stop()` function and I added some ``sdw_bread_no_pm_unlocked(&cdns->bus, 1, SDW_SCP_STAT);`` in  `sdw_cdns_clock_stop()` to see if I can find the exactly command that cause the problem. The test is still on going hope I can get the result in 24 hr.

**Comment by bardliao** [comment]
The `-61` error happened after [cdns_config_update](https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/cadence_master.c#L1446). I am wondering if any cdns configuration is different between error and normal cases. 
I added cdns register 0x0 to 0x1c dump before [cdns_config_update](https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/cadence_master.c#L1446). Hope I can see something in the test.

**Comment by bardliao** [comment]
I can see `CDNS_MCP_CONTROL` (0x4) is `0x711` before `cdns_config_update` when the `-61` issue happened.  `CDNS_MCP_CONTROL` will be `0x701` in the normal cases. I.e. `CDNS_MCP_CONTROL_HW_RST` is set before `cdns_config_update` which can explain why the codec loses sync.

**Comment by bardliao** [comment]
@plbossart A potential fix is https://github.com/thesofproject/linux/pull/2694. The test is on going.

**Comment by bardliao** [comment]
#2694 didn't solve the issue. The error became
```
[10853.235576] intel-sdw intel-sdw.2: cdns HW reset not finished yet -110
[10853.235599] intel-sdw intel-sdw.2: cannot enable clock stop on suspend
```
[dmesg_0117.txt](https://github.com/thesofproject/linux/files/5825948/dmesg_0117.txt)

I am testing with below patch now, but I am not sure if it is the right way to fix the issue.
```patch
@@ -1448,6 +1447,9 @@ int sdw_cdns_clock_stop(struct sdw_cdns *cdns, bool block_wake)
                cdns_updatel(cdns, CDNS_MCP_CONTROL,
                             CDNS_MCP_CONTROL_CMD_ACCEPT, 0);

+       cdns_updatel(cdns, CDNS_MCP_CONTROL,
+                    CDNS_MCP_CONTROL_HW_RST, 0);
+
        /* commit changes */
        ret = cdns_config_update(cdns);
        if (ret < 0) {
```

**Comment by bardliao** [comment]
> I am testing with below patch now, but I am not sure if it is the right way to fix the issue.
> 
> ```diff
> @@ -1448,6 +1447,9 @@ int sdw_cdns_clock_stop(struct sdw_cdns *cdns, bool block_wake)
>                 cdns_updatel(cdns, CDNS_MCP_CONTROL,
>                              CDNS_MCP_CONTROL_CMD_ACCEPT, 0);
> 
> +       cdns_updatel(cdns, CDNS_MCP_CONTROL,
> +                    CDNS_MCP_CONTROL_HW_RST, 0);
> +
>         /* commit changes */
>         ret = cdns_config_update(cdns);
>         if (ret < 0) {
> ```

Unfortunately this didn't work. 
```
[20574.591785] intel-sdw intel-sdw.1: cdns HW reset not finished yet -110
[20574.592836] intel-sdw intel-sdw.1: Msg ignored for Slave 1
[20574.592856] rt1308 sdw:1:25d:1308:0: Clock Stop prepare failed for slave: -61
[20574.592867] rt1308 sdw:1:25d:1308:0: pre-prepare failed:-61
[20574.592876] intel-sdw intel-sdw.1: prepare clock stop failed -61
[20574.592885] intel-sdw intel-sdw.1: cannot enable clock stop on suspend
```
Note that I removed the `return ret;` in https://github.com/thesofproject/linux/pull/2694 when I did the test.
[dmesg_0118.txt](https://github.com/thesofproject/linux/files/5828136/dmesg_0118.txt)

**Comment by bardliao** [comment]
I am running test with https://github.com/thesofproject/linux/pull/2695. Hope we can get a positive result.

**Comment by bardliao** [comment]
This happens rarely and can be recovered in the next resume. So lower the priority.

**Comment by plbossart** [comment]
@bardliao while testing PR https://github.com/thesofproject/linux/pull/2967 I found the same error with the hw_reset bits not cleared.

````
2021-06-04 02:59:44 UTC [INFO] ===== Round(456/5000) =====
2021-06-04 02:59:44 UTC [COMMAND] Run the command: rtcwake -m mem -s 5
rtcwake: assuming RTC uses UTC ...
rtcwake: wakeup from "mem" using /dev/rtc0 at Fri Jun  4 02:59:51 2021
2021-06-04 02:59:51 UTC [COMMAND] sleep for 5
2021-06-04 02:59:56 UTC [INFO] Check for the kernel log status
declare -- cmd="journalctl_cmd --since=@1622775579"
2021-06-04 02:59:56 UTC [ERROR] Caught kernel log error
===========================>>
[ 2891.854441] kernel: soundwire_intel soundwire_intel.link.1: intel_resume_runtime BUS_RESET failed: MCP_CONTROL_HW_RST is not cleared
<<===========================
````

[dmesg.txt](https://github.com/thesofproject/linux/files/6598977/dmesg.txt)

**Comment by plbossart** [comment]
This is interesting, @XiaoyunWu6666 reported a log where the problem happens on system resume in https://github.com/thesofproject/linux/issues/2943#issuecomment-856522731, while what I saw above was for pm_runtime_resume.

**Comment by bardliao** [comment]
I believe that it is a good thing that we added the "not cleared" error trace. So we can see some error that we may not notice before.

**Comment by plbossart** [comment]
@bardliao see https://github.com/thesofproject/linux/pull/2999 to refine the error trace.

**Comment by plbossart** [comment]
Interesting results at last with http://sof-ci.sh.intel.com/#/result/planresultdetail/4778?model=CML_SKU0983_SDW&testcase=check-suspend-resume-50

````
[ 4876.593489] kernel: soundwire_intel soundwire_intel.link.0: intel_resume: pm_runtime status was suspended, forcing active
[ 4876.593509] kernel: soundwire_intel soundwire_intel.link.0: intel_link_power_up: powering up all links
[ 4876.593516] kernel: soundwire_intel soundwire_intel.link.0: intel_link_power_up: first link up, programming SYNCPRD
[ 4876.593908] kernel: rt711 sdw:0:025d:0711:00: sdw_modify_slave_status: initializing enumeration and init completion for Slave 1
[ 4876.595180] kernel: soundwire_intel soundwire_intel.link.0: Slave status change: 0x2
[ 4876.595201] kernel: soundwire sdw-master-0: Slave attached, programming device number
[ 4876.595426] kernel: soundwire sdw-master-0: SDW Slave Addr: 20025d071100
[ 4876.595432] kernel: soundwire sdw-master-0: SDW Slave class_id 0x00, mfg_id 0x025d, part_id 0x0711, unique_id 0x0, version 0x2
[ 4876.595460] kernel: soundwire sdw-master-0: Slave already registered, reusing dev_num:1
[ 4876.595463] kernel: soundwire_intel soundwire_intel.link.0: intel_resume: MCP_CONTROL_HW_RST is not cleared at iteration 0
[ 4876.595984] kernel: soundwire_intel soundwire_intel.link.0: Msg ignored for Slave 0
[ 4876.596001] kernel: soundwire sdw-master-0: No more devices to enumerate
[ 4876.596088] kernel: soundwire_intel soundwire_intel.link.0: Slave status change: 0x21
[ 4876.596104] kernel: rt711 sdw:0:025d:0711:00: sdw_modify_slave_status: signaling enumeration completion for Slave 1
[ 4876.596513] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] ff01 <= 0000
[ 4876.596531] kernel: soundwire_intel soundwire_intel.link.0: intel_resume: MCP_CONTROL_HW_RST is not cleared at iteration 1
[ 4876.597321] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_read] 7520 85a0 9c20 aca0 => 00008003
[ 4876.597696] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 7520 85a0 7420 84a0 <= 001a c003
[ 4876.597800] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 3501 <= 0000
[ 4876.597903] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 3719 <= 0025
[ 4876.598007] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 3721 <= 00c0
[ 4876.598112] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 3712 <= 0020
[ 4876.598125] kernel: soundwire_intel soundwire_intel.link.0: intel_resume: MCP_CONTROL_HW_RST is not cleared at iteration 2
[ 4876.598215] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 3713 <= 0020
[ 4876.598320] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 371a <= 0020
[ 4876.598425] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 371b <= 0020
[ 4876.598632] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 7321 83a1 <= a080
[ 4876.598819] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 7321 83a1 <= 9080
[ 4876.599027] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 7308 8388 <= 6080
[ 4876.599237] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 7308 8388 <= 5080
[ 4876.599424] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 7309 8389 <= 6080
[ 4876.599521] kernel: soundwire_intel soundwire_intel.link.0: intel_resume: MCP_CONTROL_HW_RST is not cleared at iteration 3
[ 4876.599631] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 7309 8389 <= 5080
[ 4876.599738] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 4f12 <= 0091
[ 4876.599839] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 4e12 <= 00d6
[ 4876.599946] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 4d12 <= 0011
[ 4876.600047] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 4c12 <= 0020
[ 4876.600153] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 4f13 <= 0091
[ 4876.600255] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 4e13 <= 00d6
[ 4876.600362] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 4d13 <= 0011
[ 4876.600483] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 4c13 <= 0021
[ 4876.600592] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 4c21 <= 00f0
[ 4876.600693] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 4d21 <= 0011
[ 4876.600800] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 4e21 <= 0011
[ 4876.600900] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 4f21 <= 0001
[ 4876.601112] kernel: soundwire_intel soundwire_intel.link.0: intel_resume: MCP_CONTROL_HW_RST is not cleared at iteration 4
[ 4876.601255] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 7520 85a0 7420 84a0 <= 0091 0154
[ 4876.601632] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 7520 85a0 7420 84a0 <= 004a 201b
[ 4876.601985] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 7520 85a0 7420 84a0 <= 0045 5089
[ 4876.602340] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 7520 85a0 7420 84a0 <= 006b 5064
[ 4876.602529] kernel: soundwire_intel soundwire_intel.link.0: intel_resume: MCP_CONTROL_HW_RST is not cleared at iteration 5
[ 4876.602714] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 7520 85a0 7420 84a0 <= 0048 d249
[ 4876.602822] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 3501 <= 0003
[ 4876.602925] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 3501 <= 0000
[ 4876.603484] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_read] 7520 85a0 9c20 aca0 => 0000058b
[ 4876.603731] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 7520 85a0 7420 84a0 <= 006f 0580
[ 4876.604082] kernel: soundwire_intel soundwire_intel.link.0: intel_resume: MCP_CONTROL_HW_RST is not cleared at iteration 6
[ 4876.604234] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_read] 7558 85d8 9c58 acd8 => 00007888
[ 4876.604484] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_write] 7558 85d8 7458 84d8 <= 0000 f888
[ 4876.604984] kernel: rt711 sdw:0:025d:0711:00: [rt711_sdw_read] 7558 85d8 9c58 acd8 => 0000f888
[ 4876.605510] kernel: soundwire_intel soundwire_intel.link.0: intel_resume: MCP_CONTROL_HW_RST is not cleared at iteration 7
[ 4876.607114] kernel: soundwire_intel soundwire_intel.link.0: intel_resume: MCP_CONTROL_HW_RST is not cleared at iteration 8
[ 4876.608536] kernel: soundwire_intel soundwire_intel.link.0: intel_resume: MCP_CONTROL_HW_RST is not cleared at iteration 9
[ 4876.610117] kernel: soundwire_intel soundwire_intel.link.0: intel_resume failed: MCP_CONTROL_HW_RST is not cleared
````

So the HW_RST bit is never cleared, even though the control channel does work fine.

**Comment by plbossart** [comment]
@bardliao I have a possible explanation: this problem happens when the link is already pm_runtime suspended, so we are in clock-stop mode. But when we do a system suspend, we don't resume the link out of clock stop.

And when we do a system resume, we may be in a zombie state. It could very well be that we have to do a pm_runtime resume first, and do a full system suspend.

**Comment by plbossart** [comment]
See https://github.com/thesofproject/linux/pull/3006 as a tentative proposal.

**Comment by bardliao** [comment]
@plbossart The explanation makes sense and #3006 looks good to me. I started an internal test 4793 to see if we can reproduce the HW_RST bit never cleared issue.

**Comment by plbossart** [comment]
@bardliao seems like my patch creates a timeout issue for RT711. I probably forgot to reset something.

````
[ 1044.650815] kernel: soundwire_intel soundwire_intel.link.0: IO transfer timed out, cmd 3 device 1 addr b921 len 1
[ 1044.650823] kernel: soundwire sdw-master-0: trf on Slave 1 failed:-110 write addr b921 count 0
[ 1044.650907] kernel: IO error in rt711_jack_detect_handler, ret -110
````

**Comment by plbossart** [comment]
@bardliao @ranj063 I found the issue, my idea of resuming at the link level was flawed. The system suspend starts at the codec driver. Resuming at the link level resulted in an initialization of the devices and scheduling the jack detection workqueue, but the codec suspend was not called. Or rather it was already called before so all the regular suspend activity did not occur.

The right way to bring devices to full power is to call pm_runtime_resume() from the codec driver system suspend callback. That will force the parent to wake up, restart the link and hence exit clock stop...

**Comment by plbossart** [comment]
let's close this issue, we haven't seen any problems except on CML_RVP so far, and we already track clock stop issues in https://github.com/thesofproject/linux/issues/3012
